src/mlpack/methods/ann/gan/gan_impl.hpp

```diff
@@ -0,0 +1,523 @@
+/**
+ * @file gan_impl.hpp
+ * @author Kris Singh
+ * @author Shikhar Jaiswal
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license. You should have received a copy of the
+ * 3-clause BSD license along with mlpack. If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_GAN_GAN_IMPL_HPP
+#define MLPACK_METHODS_ANN_GAN_GAN_IMPL_HPP
+
+#include "gan.hpp"
+
+#include <mlpack/core.hpp>
+
+#include <mlpack/methods/ann/ffn.hpp>
+#include <mlpack/methods/ann/init_rules/network_init.hpp>
+#include <mlpack/methods/ann/visitor/output_parameter_visitor.hpp>
+#include <mlpack/methods/ann/activation_functions/softplus_function.hpp>
+#include <boost/serialization/variant.hpp>
+
+namespace mlpack {
+namespace ann /** Artifical Neural Network.  */ {
+template<
+  typename Model,
+  typename InitializationRuleType,
+  typename Noise,
+  typename PolicyType
+>
+GAN<Model, InitializationRuleType, Noise, PolicyType>::GAN(
+    arma::mat& predictors,
+    Model& generator,
+    Model& discriminator,
+    InitializationRuleType& initializeRule,
+    Noise& noiseFunction,
+    const size_t noiseDim,
+    const size_t batchSize,
+    const size_t generatorUpdateStep,
+    const size_t preTrainSize,
+    const double multiplier,
+    const double clippingParameter,
+    const double lambda):
+    predictors(predictors),
+    generator(generator),
+    discriminator(discriminator),
+    initializeRule(initializeRule),
+    noiseFunction(noiseFunction),
+    noiseDim(noiseDim),
+    batchSize(batchSize),
+    generatorUpdateStep(generatorUpdateStep),
+    preTrainSize(preTrainSize),
+    multiplier(multiplier),
+    clippingParameter(clippingParameter),
+    lambda(lambda),
+    reset(false)
+{
+  // Insert IdentityLayer for joining the Generator and Discriminator.
+  discriminator.network.insert(
+      discriminator.network.begin(),
+      new IdentityLayer<>());
+
+  counter = 0;
+  currentBatch = 0;
+
+  discriminator.deterministic = generator.deterministic = true;
```

Comment:
We should use ``this.discriminator`` or change the parameter name of to ``discriminatorModel`` or something like, that should fix the issue we have seen before.